### PR TITLE
Feature/tapestry charts

### PR DESCRIPTION
**Pull request by acron0**
Adding the heatmap feature to hecuba.

### CODE REVIEW

File: src/cljs/kixi/hecuba/tabs/hierarchy/property_details.cljs [hunk 1]
```diff
@@ -502,6 +504,15 @@
...
                 [:a {:onClick (fn [_ _] (om/update! properties :active-tab :sensors))}
                  "Sensor Charts"]]
+               [:li {:class (if (= active-tab :tapestry) "active" nil)}
+                [:a {:onClick (fn [_ _]
+                                (om/update! properties [:raw-data :sensors] (into [] (data/fetch-sensors property-id @properties)))
+                                (om/update! properties :active-tab :tapestry))}
+                 "Tapestry Charts"]]
+               (comment [:li {:class (if (= active-tab :xyplot) "active" nil)}
+                         [:a {:onClick (fn [_ _]
+                                         (om/update! properties :active-tab :xyplot))}
```
Comment (sw1nn) [inline src/cljs/kixi/hecuba/tabs/hierarchy/property_details.cljs]: should you have a `(.preventDefault event)` here ?

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs [hunk 2]
```diff
@@ -0,0 +1,385 @@
+(ns kixi.hecuba.tabs.hierarchy.tapestry
+  (:require-macros [cljs.core.async.macros :refer [go go-loop]]
+                   [hiccups.core :as hiccups])
+  (:require [clojure.string :as str]
```
Comment (sw1nn) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: `M-x cljr-sort-ns` - makes it easier to find things.

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+         shape #(conj (svg/polygon [%1 %2 %3 %4] {:fill %5}) [:title %6])}
+    :as d-spec}]
+  (let [scale-x   (:scale x-axis)
+        scale-y   (:scale y-axis)
+        pmax      (dec (count palette))
+        scale-v   (palette-scale value-domain [0 pmax])]
+    (svg/group
+     attribs
+     (for [p (nd/position-seq matrix)
+           :let [[y x] p
```
Comment (sw1nn) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: Why is this `y x` and not `x y` - looks wierd, is a comment required?

File: src/cljs/kixi/hecuba/tabs/hierarchy/property_details.cljs (hunk 1 above)
Comment (acron0) [inline src/cljs/kixi/hecuba/tabs/hierarchy/property_details.cljs]: Not necessary for `a` elements.

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+
+(defn int-to-tod
+  [num]
+  (let [hrs (mod num 24)
+        mins 0]
+    (gstring/format "%02d:%02d" hrs mins)))
+
+(defn set-new-heatmap-data!
+  [cursor data lcb ucb gradations dates]
+  (let [lcb (if (nil? lcb) (.floor js/Math (apply min data)) lcb)
```
Comment (sw1nn) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: `(if (nil? x) a b)` is equivalent to `(if-not x b a)` - kibit should show this.

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs (hunk 3 above)
Comment (acron0) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: It comes out of thi.ng that way.

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        ucb (if (nil? ucb) (.ceil js/Math (apply max data)) ucb)
+        midcb ((interpolate lcb ucb) 0.5)
+        gradations (if (nil? gradations) default-gradations gradations)
```
Comment (sw1nn) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: it's not idiomatic to use `nil?` like this if an if predicate. see earlier comment.

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs (hunk 4 above)
Comment (acron0) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: Do you mean `(if-not x a b)` ?

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+                                                  :visible false})
+                                        :y-axis (viz/linear-axis
+                                                 {:domain [0 gradations]
+                                                  :range [(- @chart-height 10) 35]
+                                                  :major 1
+                                                  :major-size 0
+                                                  :pos 10
+                                                  :label-dist -35
+                                                  :label {:text-anchor "start"}
+                                                  :format #(cond
```
Comment (sw1nn) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: check out `condp`:
```
#(condp = %
       0                         lcb
        gradations         ucb
        (/ gradations 2) midcb)
```

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn matrix [r c data]
+  (map-indexed
+   (fn[i _]
+     ((comp (partial take c) (partial drop (* c i))) data)) (range r)))
+
+(defn- data-loop [cursor input-chan]
+  (go (loop [{:keys [data dates lcb ucb grads]} (<! input-chan)]
```
Comment (sw1nn) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: there's a `go-loop` macro that does this - use that instead.

**Comment by otfrom** [comment]
The failing travis tests need to be fixed too.

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+  "This will currently only work for hourly data"
+  [data]
+  ;; step 1
+  (let [remove-times  (map #(-> %
+                                (assoc-in [:date] (.slice (:timestamp %) 0 10))
+                                (assoc-in [:time] (.slice (:timestamp %) 11 19))
+                                (dissoc :sensor_id)
+                                (dissoc :timestamp)) data)
+        grouped-data  (group-by :date remove-times)]
+    (if-not (= (count grouped-data) x-readings) ;; ensure there are 2 weeks worth
```
Comment (sw1nn) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: i'd prefer  `(if (not= a b)` over `(if-not (= a b))` - 'cos the predicate is clearer then

File: src/cljs/kixi/hecuba/widgets/datetimepicker_small.cljs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+              :on-change #(om/set-state! owner key (.. %1 -target -value))
+              :class "form-control"
+              :id id-for
+              :type "date"}]]))
+
+(defmulti calculate-new-date (fn [event _] event))
+
+(defmethod calculate-new-date :day-minus [event current-date]
+  (let [d (tf/parse current-date)]
+    (tf/unparse (tf/formatter "yyyy-MM-dd") (t/minus d (t/days 1)))))
```
Comment (sw1nn) [inline src/cljs/kixi/hecuba/widgets/datetimepicker_small.cljs]: define the formatter in a  constant?

File: src/cljs/kixi/hecuba/widgets/datetimepicker_small.cljs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+             :on-click (fn [_] (when-not inactive?
+                                 (put! event-chan {:event :date-selected :value (calculate-new-date :week-plus date)})))}]]
+
+          [:div {:style {:display "inline-block"}}
+           [:button.btn.btn-primary
+            {:type "button"
+             :id "raw-data-get-dates"
+             :style {:margin-top "10px"}
+             :on-click (fn [e _]
+                         (if-not (nil? date)
```
Comment (sw1nn) [inline src/cljs/kixi/hecuba/widgets/datetimepicker_small.cljs]: use `when date` for single branch

**Comment by sw1nn** [comment]
General comment is that nil punning is a thing that can make the code clearer.....

run kibit /eastwood on the code and I think it will highlight these things.

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs (hunk 2 above)
Comment (acron0) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: This does weird things to the require-macros key, so perhaps not 100% CLJS proof.

**Comment by acron0** [comment]
I did run kibit and the only comments are on code that isn't mine. Not tried eastwood.

**Comment by acron0** [comment]
@otfrom Have you looked at the test failure? It's pretty cryptic and could be a side-effect of the 1.7 upgrade. Anything like this you've seen before?

File: fake-data/data1.json [hunk 11]
```diff
@@ -0,0 +1 @@
...0:00.000Z","value":17.302083333333332},{"timestamp":"1970-01-14T06:00:00.000Z","value":17.3125},{"timestamp":"1970-01-14T07:00:00.000Z","value":17.322916666666668},{"timestamp":"1970-01-14T08:00:00.000Z","value":17.333333333333332},{"timestamp":"1970-01-14T09:00:00.000Z","value":17.34375},{"timestamp":"1970-01-14T10:00:00.000Z","value":17.354166666666668},{"timestamp":"1970-01-14T11:00:00.000Z","value":17.364583333333332},{"timestamp":"1970-01-14T12:00:00.000Z","value":17.375},{"timestamp":"1970-01-14T13:00:00.000Z","value":17.385416666666668},{"timestamp":"1970-01-14T14:00:00.000Z","value":17.395833333333332},{"timestamp":"1970-01-14T15:00:00.000Z","value":17.40625},{"timestamp":"1970-01-14T16:00:00.000Z","value":17.416666666666668},{"timestamp":"1970-01-14T17:00:00.000Z","value":17.427083333333332},{"timestamp":"1970-01-14T18:00:00.000Z","value":17.4375},{"timestamp":"1970-01-14T19:00:00.000Z","value":17.447916666666668},{"timestamp":"1970-01-14T20:00:00.000Z","value":17.458333333333332},{"timestamp":"1970-01-14T21:00:00.000Z","value":17.46875},{"timestamp":"1970-01-14T22:00:00.000Z","value":17.479166666666668},{"timestamp":"1970-01-14T23:00:00.000Z","value":17.489583333333332}]}
```
Comment (otfrom) [inline fake-data/data1.json]: still needed?

File: fake-data/data2.json [hunk 12]
```diff
@@ -0,0 +1 @@
...temperature","timestamp":"2012-05-13T11:00:00+0000","value":"13.893135"},{"sensor_id":"temperature","timestamp":"2012-05-13T12:00:00+0000","value":"15.30972"},{"sensor_id":"temperature","timestamp":"2012-05-13T13:00:00+0000","value":"16.184177"},{"sensor_id":"temperature","timestamp":"2012-05-13T14:00:00+0000","value":"17.01356"},{"sensor_id":"temperature","timestamp":"2012-05-13T15:00:00+0000","value":"17.449013"},{"sensor_id":"temperature","timestamp":"2012-05-13T16:00:00+0000","value":"19.470005"},{"sensor_id":"temperature","timestamp":"2012-05-13T17:00:00+0000","value":"18.438578"},{"sensor_id":"temperature","timestamp":"2012-05-13T18:00:00+0000","value":"15.559541"},{"sensor_id":"temperature","timestamp":"2012-05-13T19:00:00+0000","value":"14.128673"},{"sensor_id":"temperature","timestamp":"2012-05-13T20:00:00+0000","value":"12.893963"},{"sensor_id":"temperature","timestamp":"2012-05-13T21:00:00+0000","value":"11.31248"},{"sensor_id":"temperature","timestamp":"2012-05-13T22:00:00+0000","value":"10.094543"},{"sensor_id":"temperature","timestamp":"2012-05-13T23:00:00+0000","value":"9.210234"},{"sensor_id":"temperature","timestamp":"2012-05-14T00:00:00+0000","value":"8.333689"}]}
```
Comment (otfrom) [inline fake-data/data2.json]: still needed?

File: fake-data/serve.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
+import SimpleHTTPServer
```
Comment (otfrom) [inline fake-data/serve.py]: still needed?

File: src/clj/kixi/hecuba/api/measurements.clj [hunk 14]
```diff
@@ -156,6 +156,7 @@
         {:keys [start-date end-date device_id type]} (:items ctx)
         sensor_id (-> (:sensor ctx) :sensor_id)]
     (let [measurements (measurements/retrieve-measurements db-session start-date end-date device_id sensor_id)]
+      (println ctx)
```
Comment (otfrom) [inline src/clj/kixi/hecuba/api/measurements.clj]: remove. This will fill up the screen and logs

File: src/cljs/kixi/hecuba/model.cljs [hunk 15]
```diff
@@ -90,6 +90,11 @@
                             :sensors []
                             :selected nil
                             :date nil}
+                 :tapestry {:heatmap {:data nil}
+                            :heatmap-controls {:lcb    {:default 10}
+                                               :ucb    {:default 30}
+                                               :grads  {:default 20}}}
+                 :xyplot {}
```
Comment (otfrom) [inline src/cljs/kixi/hecuba/model.cljs]: I suppose :xyplot isn't needed quite yet.

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+(def chart-width (atom 800))
+(def chart-height (atom 600))
+(def chart-data-chan (atom nil))
+(def chart-fill-oor-cells? (atom true))
+(def default-gradations 20)
+(def invalid-cell-colour "#555555")
+
+(def colour-scheme
+  "http://colorbrewer2.org/?type=diverging&scheme=RdYlBu&n=10"
+  [[49 54 149] [69 117 180] [116 173 209] [171 217 233] [224 243 248] [254 224 144] [253 174 97] [244 109 67] [215 48 39] [165 0 38]])
```
Comment (otfrom) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: I think this is available in https://github.com/thi-ng/color but this works for here.

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs [hunk 17]
```diff
@@ -0,0 +1,382 @@
...
+          :keywords? true})))
+
+(defn update-chart-settings
+  [owner cursor heatmap-data-chan]
+  (let [raw-lcb   (read-string (.-value (om/get-node owner "lcb-input")))
+        raw-ucb   (read-string (.-value (om/get-node owner "ucb-input")))
+        raw-grads nil ;; (read-string (.-value (om/get-node owner "grads-input")))
+        lcb       (if (nil? raw-lcb) (-> cursor :heatmap-controls :lcb :default) raw-lcb)
+        ucb       (if (nil? raw-ucb) (-> cursor :heatmap-controls :ucb :default) raw-ucb)
+        grads     (if (nil? raw-grads) (-> cursor :heatmap-controls :grads :default) raw-grads)]
```
Comment (otfrom) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: see other comments on nil punning and if/if-not

File: src/cljs/kixi/hecuba/tabs/hierarchy/xyplot.cljs [hunk 18]
```diff
@@ -0,0 +1,27 @@
+(ns kixi.hecuba.tabs.hierarchy.xyplot
```
Comment (otfrom) [inline src/cljs/kixi/hecuba/tabs/hierarchy/xyplot.cljs]: save for next PR?

File: src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs (hunk 16 above)
Comment (acron0) [inline src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs]: Not this format, unfortunately.

**Comment by otfrom** [comment]
:shipit:

**Pull request merged**